batched/dense/impl/KokkosBatched_Gbtrs_Serial_Impl.hpp

```diff
@@ -0,0 +1,108 @@
+KOKKOS_INLINE_FUNCTION static int checkGbtrsInput([[maybe_unused]] const AViewType &A,
+                                                  [[maybe_unused]] const PivViewType &ipiv,
+                                                  [[maybe_unused]] const BViewType &b, [[maybe_unused]] const int kl,
+                                                  [[maybe_unused]] const int ku) {
+  static_assert(Kokkos::is_view_v<AViewType>, "KokkosBatched::gbtrs: AViewType is not a Kokkos::View.");
+  static_assert(Kokkos::is_view_v<PivViewType>, "KokkosBatched::gbtrs: PivViewType is not a Kokkos::View.");
+  static_assert(Kokkos::is_view_v<BViewType>, "KokkosBatched::gbtrs: BViewType is not a Kokkos::View.");
+  static_assert(AViewType::rank == 2, "KokkosBatched::gbtrs: AViewType must have rank 2.");
+  static_assert(PivViewType::rank == 1, "KokkosBatched::gbtrs: PivViewType must have rank 1.");
+  static_assert(BViewType::rank == 1, "KokkosBatched::gbtrs: BViewType must have rank 1.");
+  using PivValueType = typename PivViewType::non_const_value_type;
+  static_assert(std::is_integral_v<PivValueType>,
+                "KokkosBatched::gbtrs: value type of PivViewType must be an integral type.");
+#if (KOKKOSKERNELS_DEBUG_LEVEL > 0)
+  if (kl < 0) {
+    Kokkos::printf(
+        "KokkosBatched::gbtrs: input parameter kl must not be less than 0: kl "
+        "= "
+        "%d\n",
+        kl);
+    return 1;
+  }
+
+  if (ku < 0) {
+    Kokkos::printf(
+        "KokkosBatched::gbtrs: input parameter ku must not be less than 0: ku "
+        "= "
+        "%d\n",
+        ku);
+    return 1;
+  }
+
+  const int lda = A.extent(0), n = A.extent(1);
+  if (lda < (2 * kl + ku + 1)) {
+    Kokkos::printf(
+        "KokkosBatched::gbtrs: leading dimension of A must be smaller than 2 * "
+        "kl + ku + 1: "
+        "lda = %d, kl = %d, ku = %d\n",
+        lda, kl, ku);
+    return 1;
+  }
+
+  const int ldb = b.extent(0);
+  if (ldb < Kokkos::max(1, n)) {
+    Kokkos::printf(
+        "KokkosBatched::gbtrs: leading dimension of b must be smaller than "
+        "max(1, n): "
+        "ldb = %d, n = %d\n",
+        ldb, n);
+    return 1;
+  }
+
+  const int npiv = ipiv.extent(0);
+  if (npiv != n) {
+    Kokkos::printf(
+        "KokkosBatched::gbtrs: the dimension of the ipiv array must "
+        "satisfy ipiv.extent(0) == n: ipiv: %d, n: %d\n",
+        npiv, n);
+    return 1;
+  }
+
+#endif
+  return 0;
+}
+}  // namespace Impl
+
+template <typename ArgTrans>
+struct SerialGbtrs<ArgTrans, Algo::Gbtrs::Unblocked> {
+  template <typename AViewType, typename PivViewType, typename BViewType>
+  KOKKOS_INLINE_FUNCTION static int invoke(const AViewType &A, const PivViewType &piv, const BViewType &b, const int kl,
+                                           const int ku) {
+    // quick return if possible
+    if (A.extent(1) == 0) return 0;
+    auto info = Impl::checkGbtrsInput(A, piv, b, kl, ku);
```

Comment:
I would run the checks before the quick return as this might reveal a mismatch in argument sizes and inform the user on a potential issue that will otherwise be skipped...

Reply:
Actually, there is one check from lapack implementation that can be violated for `n == 0` case, assuming b size is also `0 (ldb == 0)` . If we think `n == 0` is a valid input, then we also need to modify this check.

```C++
const int ldb = b.extent(0);
if (ldb < Kokkos::max(1, n)) {
  Kokkos::printf(
      "KokkosBatched::gbtrs: leading dimension of b must not be smaller than "
      "max(1, n): "
      "ldb = %d, n = %d\n",
      ldb, n);
  return 1;
}
```

Reply:
We can modify this check to

```C++
const int ldb = b.extent(0);
if (ldb < Kokkos::max(0, n)) {
  Kokkos::printf(
      "KokkosBatched::gbtrs: leading dimension of b must not be smaller than "
      "max(0, n): "
      "ldb = %d, n = %d\n",
      ldb, n);
  return 1;
}
```